### PR TITLE
Add various tests for error event specialness

### PR DESCRIPTION
**Pull request by domenic**
Follows https://github.com/whatwg/html/pull/2398; see also https://github.com/whatwg/html/issues/2296.

### CODE REVIEW

**Comment by w3c-bots** [comment]
These tests are now available on [w3c-test.org](https://w3c-test.org/submissions/5014)

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/205148903)

# Firefox (nightly channel) #
Testing web-platform-tests at revision db4669da9b14a80f981c56a3b63210d5dbd737b9
Using browser at version BuildID 20170223110219; SourceStamp 32dcdde1fc64fc39a9065dc4218265dbc727673f
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>7 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5014/html/webappapis/scripting/events/event-handler-processing-algorithm-error/window-runtime-error.html">/html/webappapis/scripting/events/event-handler-processing-algorithm-error/window-runtime-error.html</a></summary>


|                                         Subtest                                         | Results |
|-----------------------------------------------------------------------------------------|---------|
|                                                                                         | OK      |
| `error event is weird (return true cancels; many args) on Window, with a runtime error` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5014/html/webappapis/scripting/events/event-handler-processing-algorithm-error/window-synthetic-errorevent.html">/html/webappapis/scripting/events/event-handler-processing-algorithm-error/window-synthetic-errorevent.html</a></summary>


|                                            Subtest                                             | Results |
|------------------------------------------------------------------------------------------------|---------|
|                                                                                                | OK      |
| `error event is weird (return true cancels; many args) on Window, with a synthetic ErrorEvent` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5014/html/webappapis/scripting/events/event-handler-processing-algorithm-error/window-synthetic-event.html">/html/webappapis/scripting/events/event-handler-processing-algorithm-error/window-synthetic-event.html</a></summary>


|                                             Subtest                                              | Results |
|--------------------------------------------------------------------------------------------------|---------|
|                                                                                                  | OK      |
| `error event is normal (return true does not cancel; one arg) on Window, with a synthetic Event` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5014/html/webappapis/scripting/events/event-handler-processing-algorithm-error/script-element.html">/html/webappapis/scripting/events/event-handler-processing-algorithm-error/script-element.html</a></summary>


|                                                        Subtest                                                         | Results |
|------------------------------------------------------------------------------------------------------------------------|---------|
|                                                                                                                        | OK      |
| `error event behaves normally (return true does not cancel; one arg) on a script element, with a synthetic ErrorEvent` | FAIL    |
| `error event behaves normally (return true does not cancel; one arg) on a script element, with a 404 error`            | PASS    |
| `error event behaves normally (return true does not cancel; one arg) on a script element, with a synthetic Event`      | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5014/html/webappapis/scripting/events/event-handler-processing-algorithm-error/document-synthetic-errorevent.html">/html/webappapis/scripting/events/event-handler-processing-algorithm-error/document-synthetic-errorevent.html</a></summary>


|                                                 Subtest                                                 | Results |
|---------------------------------------------------------------------------------------------------------|---------|
|                                                                                                         | OK      |
| `error event is normal (return true does not cancel; one arg) on Document, with a synthetic ErrorEvent` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5014/html/webappapis/scripting/events/event-handler-processing-algorithm-error/document-synthetic-event.html">/html/webappapis/scripting/events/event-handler-processing-algorithm-error/document-synthetic-event.html</a></summary>


|                                              Subtest                                               | Results |
|----------------------------------------------------------------------------------------------------|---------|
|                                                                                                    | OK      |
| `error event is normal (return true does not cancel; one arg) on Document, with a synthetic Event` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5014/html/webappapis/scripting/events/event-handler-processing-algorithm-error/worker.html">/html/webappapis/scripting/events/event-handler-processing-algorithm-error/worker.html</a></summary>


|                                                     Subtest                                                      | Results |
|------------------------------------------------------------------------------------------------------------------|---------|
|                                                                                                                  | OK      |
| `error event is normal (return true does not cancel; one arg) on Worker, with a synthetic Event`                 | FAIL    |
| `error event is normal (return true does not cancel; one arg) on Worker, with a synthetic ErrorEvent`            | FAIL    |
| `error event is normal (return true does not cancel; one arg) on Worker, with a syntax error in the worker code` | FAIL    |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/205148904)

# Chrome (unstable channel) #
Testing web-platform-tests at revision db4669da9b14a80f981c56a3b63210d5dbd737b9
Using browser at version 58.0.3018.3 dev
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>7 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5014/html/webappapis/scripting/events/event-handler-processing-algorithm-error/window-runtime-error.html">/html/webappapis/scripting/events/event-handler-processing-algorithm-error/window-runtime-error.html</a></summary>


|                                         Subtest                                         | Results |
|-----------------------------------------------------------------------------------------|---------|
|                                                                                         | OK      |
| `error event is weird (return true cancels; many args) on Window, with a runtime error` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5014/html/webappapis/scripting/events/event-handler-processing-algorithm-error/window-synthetic-errorevent.html">/html/webappapis/scripting/events/event-handler-processing-algorithm-error/window-synthetic-errorevent.html</a></summary>


|                                            Subtest                                             | Results |
|------------------------------------------------------------------------------------------------|---------|
|                                                                                                | OK      |
| `error event is weird (return true cancels; many args) on Window, with a synthetic ErrorEvent` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5014/html/webappapis/scripting/events/event-handler-processing-algorithm-error/window-synthetic-event.html">/html/webappapis/scripting/events/event-handler-processing-algorithm-error/window-synthetic-event.html</a></summary>


|                                             Subtest                                              | Results |
|--------------------------------------------------------------------------------------------------|---------|
|                                                                                                  | OK      |
| `error event is normal (return true does not cancel; one arg) on Window, with a synthetic Event` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5014/html/webappapis/scripting/events/event-handler-processing-algorithm-error/script-element.html">/html/webappapis/scripting/events/event-handler-processing-algorithm-error/script-element.html</a></summary>


|                                                        Subtest                                                         | Results |
|------------------------------------------------------------------------------------------------------------------------|---------|
|                                                                                                                        | OK      |
| `error event behaves normally (return true does not cancel; one arg) on a script element, with a synthetic ErrorEvent` | PASS    |
| `error event behaves normally (return true does not cancel; one arg) on a script element, with a 404 error`            | PASS    |
| `error event behaves normally (return true does not cancel; one arg) on a script element, with a synthetic Event`      | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5014/html/webappapis/scripting/events/event-handler-processing-algorithm-error/document-synthetic-errorevent.html">/html/webappapis/scripting/events/event-handler-processing-algorithm-error/document-synthetic-errorevent.html</a></summary>


|                                                 Subtest                                                 | Results |
|---------------------------------------------------------------------------------------------------------|---------|
|                                                                                                         | OK      |
| `error event is normal (return true does not cancel; one arg) on Document, with a synthetic ErrorEvent` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5014/html/webappapis/scripting/events/event-handler-processing-algorithm-error/document-synthetic-event.html">/html/webappapis/scripting/events/event-handler-processing-algorithm-error/document-synthetic-event.html</a></summary>


|                                              Subtest                                               | Results |
|----------------------------------------------------------------------------------------------------|---------|
|                                                                                                    | OK      |
| `error event is normal (return true does not cancel; one arg) on Document, with a synthetic Event` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5014/html/webappapis/scripting/events/event-handler-processing-algorithm-error/worker.html">/html/webappapis/scripting/events/event-handler-processing-algorithm-error/worker.html</a></summary>


|                                                     Subtest                                                      | Results |
|------------------------------------------------------------------------------------------------------------------|---------|
|                                                                                                                  | OK      |
| `error event is normal (return true does not cancel; one arg) on Worker, with a synthetic Event`                 | PASS    |
| `error event is normal (return true does not cancel; one arg) on Worker, with a synthetic ErrorEvent`            | PASS    |
| `error event is normal (return true does not cancel; one arg) on Worker, with a syntax error in the worker code` | PASS    |

</details>
</details>

**Comment by foolip** [comment]
One failing test for Chrome here, a bunch for Firefox. Won't test further, too time consuming, will wait for the improved tooling to materialize.

File: html/webappapis/scripting/events/event-handler-processing-algorithm-error/window-runtime-error.html [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+"use strict";
+setup({ allow_uncaught_exception: true });
+
+promise_test(t => {
+  window.onerror = t.step_func((...args) => {
+    assert_greater_than(args.length, 1);
+    return true;
+  });
+
+  const eventWatcher = new EventWatcher(t, window, "error");
```
Comment (foolip) [inline html/webappapis/scripting/events/event-handler-processing-algorithm-error/window-runtime-error.html]: Regarding my comments in https://github.com/whatwg/html/pull/2398 about synthetic events, this with one other type would be the test for that behavior.

**Review by foolip — APPROVED** [review]
Quite comprehensive, changes possibly needed depending on discussion on spec side.

File: html/webappapis/scripting/events/event-handler-processing-algorithm-error/window-runtime-error.html [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+<div id="log"></div>
+
+<script>
+"use strict";
+setup({ allow_uncaught_exception: true });
+
+promise_test(t => {
+  window.onerror = t.step_func((...args) => {
+    assert_greater_than(args.length, 1);
```
Comment (domenic) [inline html/webappapis/scripting/events/event-handler-processing-algorithm-error/window-runtime-error.html]: Good point, I don't think there is one, or at least not an easy-to-find one. I'll add some.

File: html/webappapis/scripting/events/event-handler-processing-algorithm-error/window-runtime-error.html (hunk 1 above)
Comment (domenic) [inline html/webappapis/scripting/events/event-handler-processing-algorithm-error/window-runtime-error.html]: Then it wouldn't trigger `.onerror = ...`, which is what this section of the spec is about.

Comment (foolip) [inline html/webappapis/scripting/events/event-handler-processing-algorithm-error/window-runtime-error.html]: Yes :)

**Review by foolip — APPROVED** [review]
New tests LGTM, but I think there's still nothing that would fail if there are 6 arguments. I can't remember if that's actually the case anywhere, but I seem to remember an oddity in the number of arguments.

**Comment by domenic** [comment]
Right, Edge has only 4. Nobody has 6. I guess we might as well add more coverage though...

**Pull request merged**